src/ops.rs

```diff
@@ -0,0 +1,228 @@
+use std::ops::{Add, Neg, Sub};
+
+use time::{Duration, OffsetDateTime};
+
+use crate::CalendarDuration;
+
+fn add_years(mut time: OffsetDateTime, years: i32) -> OffsetDateTime {
+    loop {
+        match time.replace_year(time.year() + years) {
+            Err(_) => {
+                // year replacement failed because current day does not exist
+                // in month of replaced year (i.e. February).
+                // decrement current day and try again.
+                time = time
+                    .replace_day(time.day() - 1)
+                    .expect("should be able to decrement day, in order in increment month");
+            }
+            Ok(new_time) => return new_time,
+        }
+    }
+}
+
+fn add_months(mut time: OffsetDateTime, months: i16) -> OffsetDateTime {
+    // Get integer for current month, using 0 as the base,
+    // and calculate the delta between the current month and target month
+    let curr_month_zero_base = time.month() as i16 - 1;
+    let total_month_delta = curr_month_zero_base + months;
+    // Calculate the change in years from the month delta
+    // by using the quotient of a division by 12 months.
+    let mut years_delta = total_month_delta / 12;
+    if total_month_delta.is_negative() {
+        // If the time change is negative, we need to
+        // subtract 1 from the year delta for correctness.
+        years_delta -= 1;
+    }
+    if years_delta != 0 {
+        // Adjust the year before adjusting the month, if necessary.
+        time = add_years(time, years_delta.into());
+    }
+    let target_month = if months < 0 {
+        time.month().nth_prev(months.abs() as u8)
```

Comment:
clippy [says](https://rust-lang.github.io/rust-clippy/master/index.html#/cast_abs_to_unsigned) to `months.unsigned_abs() as u8` here, to avoid mishandling `i16::MIN`. That can't happen here, but the compiler doesn't know that. Might as well help the linter clean.